cpp/src/predicates/is_sorted.cu

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/cudf.h>
+#include <utilities/cudf_utils.h>
+
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+namespace cudf {
+
+namespace experimental {
+
+namespace detail {
+
+template <bool has_nulls>
+auto is_sorted(
+    cudf::table_view const& in,
+    std::vector<order> const& column_order,
+    std::vector<null_order> const& null_precedence)
+{
+    cudaStream_t stream = 0;
+    auto in_d = table_device_view::create(in);
+    auto ineq_op = row_lexicographic_comparator<has_nulls>(
+        *in_d, *in_d,
+        rmm::device_vector<order>(column_order).data().get(),
+        rmm::device_vector<null_order>(null_precedence).data().get());
+
+    auto sorted = thrust::is_sorted(rmm::exec_policy(stream)->on(stream),
+                                    thrust::make_counting_iterator(0),
+                                    thrust::make_counting_iterator(in.num_rows()),
+                                    ineq_op);
+
+    return sorted;
+}
+
+} // namespace detail
+
+bool is_sorted(cudf::table_view const& in,
+               std::vector<order> const& column_order,
+               std::vector<null_order> const& null_precedence)
+{
+    if (in.num_columns() == 0 || in.num_rows() == 0) {
+        return true;
+    }
+
+    if (not column_order.empty()) {
```

Comment:
What if we have more or less than one column in the table and we have only one item in `column_order` and/or `null_precedence`? Should we use the same `CUDF_EXPECTS`, or assume those orders for all columns?

Reply:
I'm not following. The accepted states for `column_order` and `null_precedence`:
```
column_order.empty() or column_order.size() == in.num_columns()

null_precedence.empty() or null_precedence.size() == in.num_columns()
```

Anything else is an error.

Reply:
Is it worth adding any overload of `is_sorted` which takes a single `order` / `null_order`, and generates an appropriately sized vector, initialized to the given `order` / `null_order`?

Reply:
It's trivial to do this at the call site:

```
is_sorted(t); // assumes all columns ascending
is_sorted(t, std::vector<order>(ASCENDING, t.num_columns())); // this is the same thing as above
is_sorted(t, std::vector<order>(DESCENDING, t.num_columns()); // all columns descending
```